japicmp/src/main/java/japicmp/compat/CompatibilityChanges.java

```diff
@@ -221,6 +221,7 @@ private <T> void forAllSuperclasses(JApiClass jApiClass, Map<String, JApiClass>
 				}
 				classMap.put(foundClass.getFullyQualifiedName(), foundClass);
 			}
+			superclass.setJApiClass(foundClass);
```

Comment:
Why do we manipulate the model here? `CompatiblityChanges` is only about evaluating changes, not about changing the model. Has this assignment therefore to be moved to `JApiSuperlcass`?

Reply:
I noticed [`CompatibilityChanges` sets the corresponding class of implemented interfaces in method `checkIfSuperclassesOrInterfacesChangedIncompatible`](https://github.com/siom79/japicmp/blob/master/japicmp/src/main/java/japicmp/compat/CompatibilityChanges.java#L802) so I figured there were certain cases where `correspondingJApiClass` is not known at creation time and therefore has to be retrofitted later on by `CompatiblityChanges`.

In this case, `JApiSuperclass` is [created by `JApiClass`](https://github.com/siom79/japicmp/blob/master/japicmp/src/main/java/japicmp/model/JApiClass.java#L262):

```java
JApiSuperclass retVal = new JApiSuperclass(this, /*...*/);

```

Once `JApiSuperclass.jApiClass` has been initialized with `this`, then [`JApiSuperclass` sets `JApiSuperclass.correspondingJApiClass` to `this`](https://github.com/siom79/japicmp/blob/master/japicmp/src/main/java/japicmp/model/JApiClass.java#L296) too:

```java
retVal.setJApiClass(this);
```

I don't think that was the original intention for `correspondingJApiClass`, so I thought the assignment was wrong and that this could be one of those cases where the corresponding class needs to be initialized after the instance has been created.

However, if `CompatiblityChanges` is only about evaluating changes, we should maybe move those assignments to the point where `JApiImplementedInterface` and `JApiSuperclass` are actually instantiated. Would that make more sense?

Reply:
[Here](https://github.com/siom79/japicmp/blob/176547ce093ed8b4623bb3af68f5e8a0b6faefb0/japicmp/src/main/java/japicmp/model/JApiImplementedInterface.java#L62) the attribute `correspondingJApiClass` is evaluated on the `JApiImplementedInterface`. It was added, because the interface itself may have incompatible changes that affect the implementing class (e.g. method removed from the interface but the class still has the method implementation, see [here](https://docs.oracle.com/javase/specs/jls/se7/html/jls-13.html#jls-13.5.3)).
I think that setting this value so late is due to performance, because otherwise one would have to wait until all classes are processed, to set the corresponding JApiClass instance (as it may not be available while processing the class itself). Especially as I wanted evaluate if parallelizing the construction of the JApiClass is possible (currently it is single threaded).
So it may be ok to set the class this lately.

Reply:
> So it may be ok to set the class this lately.

Do you mean it may be ok to do that for both `JApiImplementedInterface` and `JApiSuperclass`? Or just for `JApiImplementedInterface`?